### PR TITLE
Adding homescreen widgets codelab

### PR DESCRIPTION
**Pull request by leighajarett**
This adds the homescreen widget codelab.

I have tested the iOS start and final, but I couldn't get the Android start or final to work because of some gradle issue. @ericwindmill can you verify that this is all working on Android?

Outside of @domesticmouse's review, here are the remaining TO DOs for merging this PR:

- [x] @ericwindmill to test current version start and final on Android
- [x] Break into 4 steps to match the written codelab (after Brett's review)

## Pre-launch Checklist

- [X] I read the [Effective Dart: Style] _recently_, and have followed its advice.
- [X] I signed the [CLA].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.

### CODE REVIEW

File: homescreen_codelab/step_00/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# homescreen_widgets
```
Comment (domesticmouse) [inline homescreen_codelab/step_00/README.md:1]: I frequently lose this file, as it isn't that helpful in this case.

File: homescreen_codelab/step_00/analysis_options.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# This file configures the analyzer, which statically analyzes Dart code to
```
Comment (domesticmouse) [inline homescreen_codelab/step_00/analysis_options.yaml:1]: Please replace this with the repo wide config. See https://github.com/flutter/codelabs/blob/main/adaptive_app/step_03/analysis_options.yaml for an example.

File: homescreen_codelab/step_00/android/app/src/main/java/com/leighajarett/homescreen_widgets/NewsWidget.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    override fun onUpdate(
+        context: Context,
+        appWidgetManager: AppWidgetManager,
+        appWidgetIds: IntArray,
+    ) {
+        val widgetData = HomeWidgetPlugin.getData(context)
+        for (appWidgetId in appWidgetIds) {
+            val views = RemoteViews(context.packageName, R.layout.news_widget).apply {
+
+                // Swap title text by fetching the string from local storage
```
Comment (domesticmouse) [inline homescreen_codelab/step_00/android/app/src/main/java/com/leighajarett/homescreen_widgets/NewsWidget.kt:18]: I usually migrate code comments into the codelab text. The width of the code inclusion is quite narrow, which is frequently an issue for code comments.

File: homescreen_codelab/step_00/android/app/src/main/res/drawable-v21/app_widget_background.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?><!--
+Background for widgets to make the rounded corners based on the
+appWidgetRadius attribute value
+-->
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+
+    <corners android:radius="?attr/appWidgetRadius" />
+    <solid android:color="?android:attr/colorBackground" />
+</shape>
```
Comment (domesticmouse) [inline homescreen_codelab/step_00/android/app/src/main/res/drawable-v21/app_widget_background.xml:10]: Consider adding a newline at end of file to get rid of this warning

File: homescreen_codelab/step_00/linux/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,138 @@
+# Project-level configuration.
```
Comment (domesticmouse) [inline homescreen_codelab/step_00/linux/CMakeLists.txt:1]: I believe this codelab is for Android and iOS only, please delete the `linux` directory

File: homescreen_codelab/step_00/macos/.gitignore [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Flutter-related
```
Comment (domesticmouse) [inline homescreen_codelab/step_00/macos/.gitignore:1]: I believe this codelab is for Android and iOS only, please delete the `macos` directory

File: homescreen_codelab/step_00/web/favicon.png
Comment (domesticmouse) [inline homescreen_codelab/step_00/web/favicon.png:1]: I believe this codelab is for Android and iOS only, please delete the `web` directory

File: homescreen_codelab/step_00/windows/.gitignore [hunk 7]
```diff
@@ -0,0 +1,17 @@
+flutter/ephemeral/
```
Comment (domesticmouse) [inline homescreen_codelab/step_00/windows/.gitignore:1]: I believe this codelab is for Android and iOS only, please delete the `windows` directory

**Review by domesticmouse — COMMENTED** [review]
FIrst pass, just reviewing `step_00`.

File: homescreen_codelab/step_00/android/build.gradle [hunk 8]
```diff
@@ -0,0 +1,31 @@
+buildscript {
+    ext.kotlin_version = '1.6.10'
```
Comment (ericwindmill) [inline homescreen_codelab/step_00/android/build.gradle:2]: @ericwindmill change to 1.8.0 everywhere

File: homescreen_codelab/step_00/README.md (hunk 1 above)
Comment (leighajarett) [reply homescreen_codelab/step_00/README.md:1]: SG!

File: homescreen_codelab/step_00/analysis_options.yaml (hunk 2 above)
Comment (leighajarett) [reply homescreen_codelab/step_00/analysis_options.yaml:1]: Done

File: homescreen_codelab/step_00/android/build.gradle (hunk 8 above)
Comment (leighajarett) [reply homescreen_codelab/step_00/android/build.gradle:2]: @ericwindmill are you going to make this change on the branch?

Comment (ericwindmill) [reply homescreen_codelab/step_00/android/build.gradle:2]: I fixed this

**Review by domesticmouse — APPROVED** [review]

**Comment by tashiwangdi** [comment]
@leighajarett Thank you for adding home screen widget tutorial to codelabs 💪. I have followed the steps to add the widget to home screen on iOS simulator and got the app running. However, I am not able to get the news title and description on to the widget. I cant seem to figure out what am I missing 😏 
Have you had a similar issue when you were building the app/widget? Much appreciate any input. 

Regards
Tashi

**Comment by ericwindmill** [comment]
@tashiwangdi are you seeing any errors, or is it just not rendering as you'd expect? Can you file an issue at https://github.com/flutter/flutter/issues so we can track?

**Comment by tashiwangdi** [comment]
@ericwindmill 
I don't see errors. I will re-check my configuration and may be do a clean restart again to see if I can make it work. I will file an issue if I see the issue again. Thanks.

**Pull request merged**